### PR TITLE
Fixed Issue with DCS Serializing Get-Only Collection. 

### LINKED ISSUE #22965: Exception when serialize class contains Dictionary and DateTimeOffset via DataContractSerializer
@jiemon commented on [Fri Aug 04 2017](https://github.com/dotnet/core/issues/789)

I try serialize via DataContractSerializer simple class by example:
```
public class SimpleExampleClass2
		{
			private Dictionary<string, string> _attributes1;
			public Dictionary<string, string> Attributes1 => _attributes1;
			public DateTimeOffset TestDate { get; set; }
		}
```
Serialization code:
```
var test = new SimpleExampleClass2() { TestDate = DateTimeOffset.Now };
			var ser = new DataContractSerializer(typeof(SimpleExampleClass2));
			MemoryStream memoryStream = new MemoryStream();
			ser.WriteObject(memoryStream, test);
			memoryStream.Seek(0, SeekOrigin.Begin);
			return File(memoryStream, "application/xml", $"test.xml");
```

But on line "ser.WriteObject(memoryStream, test)" I get exception:

> InvalidDataContractException: Type 'System.Runtime.Serialization.DateTimeOffsetAdapter' cannot be serialized. Consider marking it with the DataContractAttribute attribute, and marking all of its members you want serialized with the DataMemberAttribute attribute. If the type is a collection, consider marking it with the CollectionDataContractAttribute. See the Microsoft .NET Framework documentation for other supported types.

Is this bug in DataContractSerializer?

---

@Petermarcu commented on [Fri Aug 04 2017](https://github.com/dotnet/core/issues/789#issuecomment-320271980)

@zhenlan @ViktorHofer @danmosemsft 

---

@ViktorHofer commented on [Fri Aug 04 2017](https://github.com/dotnet/core/issues/789#issuecomment-320273333)

It's not related to binary serialization with BinaryFormatter.

---

@Petermarcu commented on [Fri Aug 04 2017](https://github.com/dotnet/core/issues/789#issuecomment-320274307)

@zhenlan Any thoughts?

---

@zhenlan commented on [Fri Aug 04 2017](https://github.com/dotnet/core/issues/789#issuecomment-320295733)

+ @shmao

### PR DESCRIPTION
**Pull request by shmao**
Fixed Issue with DCS Serializing Get-Only Collection. 

Fix #22965

### CODE REVIEW

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatGeneratorStatics.cs [hunk 1]
```diff
@@ -249,6 +249,20 @@ internal static MethodInfo StoreCollectionMemberInfoMethod
...
 
+        private static MethodInfo s_resetCollectionMemberInfoMethod;
+        internal static MethodInfo ResetCollectionMemberInfoMethod
+        {
+            get
+            {
+                if (s_resetCollectionMemberInfoMethod == null)
+                {
+                    s_resetCollectionMemberInfoMethod = typeof(XmlObjectSerializerReadContext).GetMethod("ResetCollectionMemberInfo", Globals.ScanAllMembers, new Type[] { });
+                    Debug.Assert(s_resetCollectionMemberInfoMethod != null);
```
Comment (mconnew) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatGeneratorStatics.cs]: Does the nameof() operator do the right thing if you specify a method in another class? If it does, I would prefer that to an inline string constant.

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatWriterGenerator.cs [hunk 2]
```diff
@@ -286,6 +286,12 @@ private int WriteMembers(ClassDataContract classContract, LocalBuilder extension
                         _ilg.Load(_contextArg);
                         _ilg.Call(XmlFormatGeneratorStatics.StoreIsGetOnlyCollectionMethod);
                     }
+                    else
+                    {
+                        _ilg.Load(_contextArg);
+                        _ilg.Call(XmlFormatGeneratorStatics.ResetIsGetOnlyCollectionMethod);
```
Comment (mconnew) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatWriterGenerator.cs]: The json generators use a different _ilg.Call overload which looks a little cleaner than this. Did you use this version to be consistent with the first statement block of the if?

**Comment by mconnew** [comment]
:shipit:

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatGeneratorStatics.cs (hunk 1 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatGeneratorStatics.cs]: Yes, using `nameof()` would be better. I was following other methods here.

**Comment by shmao** [comment]
System.Drawing.Common.Tests failed.

@dotnet-bot test Linux x64 Release Build

**Pull request merged**